### PR TITLE
[RTO-2859] Clone incompatible with filedocument

### PR DESCRIPTION
**Pull request by jds-mendix**

### CODE REVIEW

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 1]
```diff
@@ -255,6 +258,10 @@ public static String getValueOfPath(IContext context, IMendixObject substitute,
 		}
 	}
 
+	private static boolean isFileDocument(IMendixObject object) {
+	    return Core.isSubClassOf(Core.getMetaObject(FileDocument.entityName), object.getMetaObject());
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/communitycommons/ORM.java:262]: Indent is inconsistent with the rest of the file, which seems to be using tabs.

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 2]
```diff
@@ -267,6 +274,9 @@ public static Boolean cloneObject(IContext c, IMendixObject source,
 			if (m instanceof MendixAutoNumber) {
 				continue;
 			}
+			if ((isFileDocument(source) || isFileDocument(target)) && "__UUID__".equals(m.getName())) {
+			  continue;
+            }
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/communitycommons/ORM.java:279]: Indent is wrong. It seems the rest of the file is using tabs.

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 3]
```diff
@@ -267,6 +274,9 @@ public static Boolean cloneObject(IContext c, IMendixObject source,
 			if (m instanceof MendixAutoNumber) {
 				continue;
 			}
+			if ((isFileDocument(source) || isFileDocument(target)) && "__UUID__".equals(m.getName())) {
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/communitycommons/ORM.java:277]: Please swap the sides of the `&&`.

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 4]
```diff
@@ -348,6 +358,9 @@ public static void copyAttributes(IContext context, IMendixObject source, IMendi
 			if (e.isVirtual() || e.getType() == PrimitiveType.AutoNumber) {
 				continue;
 			}
+            if ((isFileDocument(source) || isFileDocument(target)) && "__UUID__".equals(e.getName())) {
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/communitycommons/ORM.java:361]: Please swap the sides of the `&&`.

File: src/CommunityCommons/javasource/myfirstmodule/actions/Test_Clone_Java.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	public Test_Clone_Java(IContext context)
+	{
+		super(context);
+	}
+
+	@java.lang.Override
+	public java.lang.Boolean executeAction() throws Exception
+	{
+		// BEGIN USER CODE
+        FileDocument doc = new FileDocument(getContext());
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/myfirstmodule/actions/Test_Clone_Java.java:29]: There's a mix of spaces and tabs for indentation, which makes it look sloppy. Please use one or the other.

**Review by hugomendix — CHANGES_REQUESTED** [review]
I've made a few minor comments. Nothing structural.

File: src/CommunityCommons/javasource/atlas_ui_resources/proxies/constants/Constants.java [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
-
-public class Constants
-{
-	// These are the constants for the Atlas_UI_Resources module
-
-	public static java.lang.String getAtlas_UI_Resources_Version()
-	{
-		return (java.lang.String)Core.getConfiguration().getConstantValue("Atlas_UI_Resources.Atlas_UI_Resources_Version");
-	}
+// This file was generated by Mendix Studio Pro.
```
[the remark this answers is not in the document]
Comment (jds-mendix) [reply src/CommunityCommons/javasource/atlas_ui_resources/proxies/constants/Constants.java:1]: I believe these are changes in whitespaces, Studio Pro generates them in this fashion, I would prefer keeping them in the way they are generated so we don't have those changes every single time we run the project.

[the remark this answers is not in the document]
Comment (hugomendix) [reply src/CommunityCommons/javasource/atlas_ui_resources/proxies/constants/Constants.java:1]: Can we put them in a separate PR then? I'll review/approve it.

[the remark this answers is not in the document]
Comment (jds-mendix) [reply src/CommunityCommons/javasource/atlas_ui_resources/proxies/constants/Constants.java:1]: They ended up here out of a `-a` habit. I'll revert

[the remark this answers is not in the document]
Comment (jds-mendix) [reply src/CommunityCommons/javasource/atlas_ui_resources/proxies/constants/Constants.java]: Done

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 7]
```diff
@@ -108,7 +109,9 @@ private static void duplicate(IContext ctx, IMendixObject src, IMendixObject tar
 					tar.setValue(ctx, key, res);
 				} else if (m instanceof MendixAutoNumber) //skip autonumbers! Ticket 14893
 				{
-					continue;
+                    // do nothing
+                } else if((isFileDocument(src) || isFileDocument(tar)) && "__UUID__".equals(key)) {
```
[the remark this answers is not in the document]
Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:113]: Good idea, done

File: src/CommunityCommons/javasource/communitycommons/ORM.java (hunk 1 above)
Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:262]: Changed.

Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:262]: I'd love that to be configurable in Studio Pro

File: src/CommunityCommons/javasource/communitycommons/ORM.java (hunk 2 above)
Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:279]: And done

File: src/CommunityCommons/javasource/communitycommons/ORM.java (hunk 3 above)
Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:277]: Done

File: src/CommunityCommons/javasource/communitycommons/ORM.java (hunk 4 above)
Comment (jds-mendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java:361]: Done

File: src/CommunityCommons/javasource/myfirstmodule/actions/Test_Clone_Java.java (hunk 5 above)
Comment (jds-mendix) [reply src/CommunityCommons/javasource/myfirstmodule/actions/Test_Clone_Java.java:29]: Fixed

File: src/CommunityCommons/javasource/communitycommons/ORM.java [hunk 8]
```diff
@@ -108,7 +109,9 @@ private static void duplicate(IContext ctx, IMendixObject src, IMendixObject tar
 					tar.setValue(ctx, key, res);
 				} else if (m instanceof MendixAutoNumber) //skip autonumbers! Ticket 14893
 				{
-					continue;
+                    // do nothing
```
Comment (hugomendix) [inline src/CommunityCommons/javasource/communitycommons/ORM.java:112]: Can you fix the indent here as well?

**Review by hugomendix — APPROVED** [review]
One last comment on indentation, but already approved.

File: src/CommunityCommons/javasource/communitycommons/ORM.java (hunk 1 above)
Comment (hugomendix) [reply src/CommunityCommons/javasource/communitycommons/ORM.java]: Fully agree!

**Pull request merged**